File: March-2023/Java/CheckCompletenessOfABinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
// 958. Check Completeness of a Binary Tree
class Solution {
    public boolean isCompleteTree(TreeNode root) {
    
        Queue<TreeNode> q = new LinkedList<>();
        boolean isnull = false;
        q.add(root);
        while(!q.isEmpty()){
            int size = q.size();
            while(size-->0){
                TreeNode t = q.poll();
                //check if node is null
                if(t==null){
                    isnull=true;
                }
                else{
                    // check if previous node is null (all nodes should be far left in a complete binary tree)
                    if(isnull==true){
                        return false;
                    }
                    q.add(t.left);
                    q.add(t.right);
                }
            }
       }
        return true;
    }
    
}

File: March-2023/C++/CheckCompletenessOfABinaryTree.cxx
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 // 958. Check Completeness of a Binary Tree
class Solution {
public:
    bool isCompleteTree(TreeNode* root) {
        queue<TreeNode*> q;
        bool isnull = false;
        q.push(root);
        while(!q.empty()){
            int size = q.size();
            while(size-->0){
                TreeNode* t = q.front();
                q.pop();
                //check if node is null
                if(t==nullptr){
                    isnull=true;
                }
                else{
                    // check if previous node is null (all nodes should be far left in a complete binary tree)
                    if(isnull==true){
                        return false;
                    }
                    q.push(t->left);
                    q.push(t->right);
                }
            }
       }
        return true;
    }
};
